autogen/cluster.tf

```diff
@@ -219,14 +219,75 @@ resource "google_container_cluster" "primary" {
 /******************************************
   Create Container Cluster node pools
  *****************************************/
+locals {
+  force_node_pool_recreation_resources = [
+    "disk_size_gb",
+    "disk_type",
+    "accelerator_count",
+    "accelerator_type",
+    "local_ssd_count",
+    "machine_type",
+    "preemptible",
+    "service_account",
+  ]
+}
+
+resource "random_id" "name" {
+  count       = var.node_pools_create_before_destroy ? length(var.node_pools) : 0
+  byte_length = 2
+  prefix      = format("%s-", lookup(var.node_pools[count.index], "name"))
+  keepers = merge(
+    zipmap(
+      local.force_node_pool_recreation_resources,
+      [for keeper in local.force_node_pool_recreation_resources : lookup(var.node_pools[count.index], keeper, "")]
+    ),
+    {
+      labels_all = join(",", keys(var.node_pools_labels["all"]), values(var.node_pools_labels["all"]))
+    },
+    {
+      labels_node_pool = join(",", keys(var.node_pools_labels[var.node_pools[count.index]["name"]]), values(var.node_pools_labels[var.node_pools[count.index]["name"]]))
+    },
+    {
+      metadata_all = join(",", keys(var.node_pools_metadata["all"]), values(var.node_pools_metadata["all"]))
+    },
+    {
+      metadata_node_pool = join(",",
+        keys(var.node_pools_metadata[var.node_pools[count.index]["name"]]),
+        values(var.node_pools_metadata[var.node_pools[count.index]["name"]])
+      )
+    },
+    {
+      oauth_scopes_all = join(",", var.node_pools_oauth_scopes["all"])
+    },
+    {
+      oauth_scopes_node_pool = join(",", var.node_pools_oauth_scopes[var.node_pools[count.index]["name"]])
+    },
+    {
+      tags_all = join(",", var.node_pools_tags["all"])
+    },
+    {
+      tags_node_pool = join(",", var.node_pools_tags[var.node_pools[count.index]["name"]])
+    }
+  )
+}
+
+{% for create_before_destroy_value in ["false", "true"] %}
```

Comment:
Is there a reason these need to be two different node pool resources? I'd like to avoid that complexity as much as possible.

Reply:
Unfortunately, the `lifecycle` block does not support interpolations [according to the docs](https://www.terraform.io/docs/configuration/resources.html#lifecycle-lifecycle-customizations):

> The `lifecycle` settings all effect how Terraform constructs and traverses the dependency graph. As a result, only literal values can be used because the processing happens too early for arbitrary expression evaluation.

Additionally, discussed further in https://github.com/hashicorp/terraform/issues/3116

As I see it, the value has to be a literal in the lifecycle block hence the templating of two different node pools. 

If you want to avoid the two different node pool resources, it sounds like there are a couple of options until the lifecycle block can use interpolations:
1 - switch to `create_before_destroy = true` for all node pools
2 - no support for `create_before_destroy = true`
3 - ???

Maybe there is another option?

Reply:
I see. In that case, I think we might want to switch to an additional (templated) submodule?

Reply:
To be clear, your suggestion is to add another `module` to the `MODULES` variable here: 
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/helpers/generate_modules/generate_modules.py#L44

Which contains another variable such as `'lifecycle_variant': True` and add some templating to `autogen/cluster.tf` based on this variable, correct?

Reply:
Created a new templated submodule, `private-cluster-lifecycle-variant`, (I'm open to other names. Not sure what to call this module(s)) which adds the `create_before_destroy = true` to the lifecycle block.

Reply:
@morgante - is `private-cluster-lifecycle-variant` along the lines of what you were suggesting?

Reply:
Yes, that's exactly what I was thinking. We might try for a more descriptive name though? `update-variant`?

Reply:
done. replaced `lifecycle-variant` with `update-variant`.